File: src/p0187/cpp/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;


class Solution {
public:
    vector<string> findRepeatedDnaSequences(const string &s) {
        vector<string> result;
        unordered_map<int, int> count;
        for (int i = 0, h = 0; i < s.size(); i++) {
            h = ((h & 0x07ffffff) << 3) | (s[i] & 7);
            if ((count[h]++) == 1) {
                result.push_back(s.substr(i - 9, 10));
            }
        }
        return result;
    }
};

File: src/p0187/java/solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        HashMap<Integer, Integer> count = new HashMap<>();
        List<String> result = new ArrayList<>();
        for (int i = 0, h = 0; i < s.length(); i++) {
            h = ((h & 0x07ffffff) << 3) | (s.charAt(i) & 7);
            int old = count.getOrDefault(h, 0);
            if (old == 1) {
                result.add(s.substring(i - 9, i + 1));
            }
            count.put(h, old + 1);
        }
        return result;
    }
}
